src/int_allocator.rs

```diff
@@ -29,62 +37,66 @@ impl IntAllocator {
     pub fn new() -> Self {
         let mut r = IntAllocator {
             u8_vec: Vec::new(),
-            node_vec: Vec::new(),
+            pair_vec: Vec::new(),
+            atom_vec: Vec::new(),
         };
         r.u8_vec.reserve(1024 * 1024);
+        r.atom_vec.reserve(256);
+        r.pair_vec.reserve(256);
         r.u8_vec.push(1_u8);
         // Preallocated empty list
-        r.node_vec
-            .push(NodePtr::Atom(IntAtomBuf { start: 0, end: 0 }));
+        r.atom_vec.push(IntAtomBuf { start: 0, end: 0 });
         // Preallocated 1
-        r.node_vec
-            .push(NodePtr::Atom(IntAtomBuf { start: 0, end: 1 }));
+        r.atom_vec.push(IntAtomBuf { start: 0, end: 1 });
         r
     }
 }
 
 impl Allocator for IntAllocator {
-    type Ptr = u32;
+    type Ptr = i32;
     type AtomBuf = IntAtomBuf;
 
     fn new_atom(&mut self, v: &[u8]) -> Self::Ptr {
         let start = self.u8_vec.len() as u32;
         self.u8_vec.extend_from_slice(v);
         let end = self.u8_vec.len() as u32;
-        let r = self.node_vec.len() as u32;
-        self.node_vec.push(NodePtr::Atom(IntAtomBuf { start, end }));
-        r
+        self.atom_vec.push(IntAtomBuf { start, end });
+        -(self.atom_vec.len() as i32)
```

Comment:
Can `self.atom_vec.len() == 0` in this scheme?

Reply:
in theory it could, but there are always two atoms that are pre-allocated. `1` and `()`, so it won't. Also, these atom-indices should only ever come back from `new_atom()`, so they should also never be out-of-range